src/python/pants/init/specs_calculator.py

```diff
@@ -69,10 +71,37 @@ def calculate_specs(
             "The `--changed-*` options are only available if Git is used for the repository."
         )
 
-    changed_files = tuple(changed_options.changed_files(maybe_git_worktree.git_worktree))
+    (files_with_sources_blocks,) = session.product_request(
+        FilesWithSourceBlocks, [Params(bootstrap_environment)]
+    )
+    changed_files = tuple(
+        file
+        for file in changed_options.changed_files(maybe_git_worktree.git_worktree)
+        # We want to exclude the file from the normal processing flow if it has associated
+        # targets with text blocks. These files are handled with special logic.
+        if file not in files_with_sources_blocks
+    )
     file_literal_specs = tuple(FileLiteralSpec(f) for f in changed_files)
 
-    changed_request = ChangedRequest(changed_files, changed_options.dependents)
+    sources_blocks = FrozenDict(
+        (
+            path,
+            # Hunk stores information about the old block and the new block.
+            # Here we only care about the final state, so we take `hunk.right`.
+            TextBlocks(hunk.right for hunk in hunks),
```

Comment:
Depending on what the block class ends up being, we prob. should filter so we don't end up with empty blocks in this list (i.e. `block.count > 0` only)